pkg/processor/runtime/python/py/_nuclio_wrapper.py

```diff
@@ -116,37 +103,58 @@ async def serve_requests(self, num_requests=None):
                 event_message_length = await self._resolve_event_message_length()
 
                 # resolve event message
-                event_message = await self._resolve_event(event_message_length)
-
-                # instantiate event message
-                event = nuclio_sdk.Event.deserialize(event_message, kind=self._event_deserializer_kind)
+                event = await self._resolve_event(event_message_length)
 
                 try:
+
+                    # handle event
                     await self._handle_event(event)
                 except BaseException as exc:
                     self._on_handle_event_error(exc)
 
             except WrapperFatalException as exc:
                 self._on_serving_error(exc)
 
-                # explode
+                # explode, unrecoverable exception
                 self._shutdown(error_code=1)
 
             except UnicodeDecodeError as exc:
 
                 # reset unpacker to avoid consecutive errors
+                # this may happen when msgpack fails to decode a non-utf8 events
                 self._unpacker = self._resolve_unpacker()
                 self._on_serving_error(exc)
 
             except Exception as exc:
                 self._on_serving_error(exc)
 
             # for testing, we can ask wrapper to only read a set number of requests
-            if num_requests is not None and num_requests != 0:
+            if num_requests is not None:
                 num_requests -= 1
+                if num_requests <= 0:
+                    break
+
+    async def initialize(self):
+
+        # call init_context
+        await self._initialize_context()
+
+        # indicate that we're ready
+        self._write_packet_to_processor('s')
 
-            if num_requests == 0:
-                break
+    async def _initialize_context(self):
+
+        # call init context
+        if hasattr(self._entrypoint_module, 'init_context'):
+            try:
+                init_context = getattr(self._entrypoint_module, 'init_context')
+                init_context_result = getattr(self._entrypoint_module, 'init_context')(self._context)
+                if asyncio.iscoroutinefunction(init_context):
```

Comment:
```suggestion
                return await asyncio.coroutine(init_context_result)()
```
no need to condition, it is safe to re-wrap a coroutine

Reply:
deprecated on 3.8